enhancements/network/ovn-kubernetes-ipsec.md

```diff
@@ -0,0 +1,298 @@
+- Will not explicitly encrypt cluster management traffic. Cluster management
+traffic has already been encrypted using TLS.
+- Will not allow granular configuration of which nodes will have IPsec enabled
+(for example Namespace to Namespace encryption) or how IPsec is configured
+between specific nodes. This enhancement will enable IPsec between all worker
+nodes within the cluster indiscriminately.
+- Will not provide IPsec connectivity between any worker node that does not form
+part of the Kubernetes cluster. This enhancement is only for Node-to-Node
+(east-west) communication within the cluster.
+
+## Proposal
+
+The following changes are proposed to the [network.operator.openshift.io API](https://github.com/openshift/api/blob/master/operator/v1/types_network.go).
+
+```go
+// ovnKubernetesConfig contains the configuration parameters for networks
+// using the ovn-kubernetes network project
+type OVNKubernetesConfig struct {
+	// mtu is the MTU to use for the tunnel interface. This must be 100
+	// bytes smaller than the uplink mtu.
+	// Default is 1400
+	// +kubebuilder:validation:Minimum=0
+	// +optional
+	MTU *uint32 `json:"mtu,omitempty"`
+	// geneve port is the UDP port to be used by geneve encapulation.
+	// Default is 6081
+	// +kubebuilder:validation:Minimum=1
+	// +optional
+	GenevePort *uint32 `json:"genevePort,omitempty"`
+	// HybridOverlayConfig configures an additional overlay network for peers that are
+	// not using OVN.
+	// +optional
+  HybridOverlayConfig *HybridOverlayConfig `json:"hybridOverlayConfig,omitempty"`
+  // EnableIPsec enables IPsec encryption on all pod-pod traffic within the cluser. If unset, the plugin will default
+  // EnableIPsec to false.
+  // +optional
+  EnableIPsec *bool `json:"enableIPsec,omitempty"`
+}
+```
+
+This will result in an updated OVNKubernetes Operator configuration object:
+(.spec.defaultNetwork.ovnKubernetesConfig) which will enable cluster-wide
+configuration of IPsec at cluster installation time:
+
+spec:
+  defaultNetwork:
+    type: OVNKubernetes
+    ovnKubernetesConfig:
+      mtu: 1400
+      genevePort: 6081
+      **enableIPsec: true/false**
+
+### User Stories
+
+#### Story 1
+
+As a cluster operator, I want to be able to enable IPsec at cluster installation
+time in the OVNKubernetes plugin in order to encrypt, authenticate and ensure
+the integrity of all pod-pod traffic.
+
+### Implementation Details/Notes/Constraints [optional]
+
+It is expected that the following changes will need to be made in order to
+enable this enhancement:
+
+- OVNKubernetes pod Dockerfiles will need to be updated in order to install
+  necessary components required by OVN/OVS to enable IPsec (e.g.
+  openvswitch-ipsec, libreswan, etc).
+- Modify Cluster Network Operator manifests to configure start up of packages
+  required by OVN/OVS in order to enable IPsec (e.g. ovs-monitor-ipsec, pluto,
+  etc) and determine how to containerize these components.
```

Comment:
When we had discussed non-OVN-based IPsec before, the consensus was that we would not want to run the ipsec daemons in containers, because if those containers were restarted (which would happen every time there was an OCP micro version upgrade) it would greatly disrupt traffic. Is that not the case with the OVN-based version?

If we need to avoid restarting the ipsec-related daemons, then we would need to run them on the host as systemd services (as we do with OVS itself now), not as containers.

Reply:
> When we had discussed non-OVN-based IPsec before, the consensus was that we would not want to run the ipsec daemons in containers, because if those containers were restarted (which would happen every time there was an OCP micro version upgrade) it would greatly disrupt traffic

With containerized pluto daemons and IKEv2 opportunistic connections, in theory there should be no tunnel disruption in this scenario, as the new post-restart IKEv2 SAs would inherit the existing kernel SAs.  Not saying that this would be a better approach, but to my understanding it could be done.

Reply:
This is a good conversation and I will have to look into this in more detail.

@mrogers950 I assume that during restart no new SAs could be established? However, I presume this is unavoidable?

@danwinship Could you provide more information as to what "greatly disrupt traffic" means? Is it, as is suggested by @mrogers950, that the existing SAs could not be inherited?

Reply:
The worry was that no traffic could be sent to or from the node while the ipsec pod was restarting.

Which, if that was true, might imply that the node could become completely isolated in some cases, if the old ipsec pod exits and the masters can't reach the node to deploy a new one.

Reply:
Or if the nodes cannot reach the registry to pull the new image.

Upgrades definitely need to be carefully thought out. We already have some infrastructure for ensuring systemd services are started based on certain configuration options - we could build off that.

Reply:
@squeed Can you point me to that?

Reply:
For sure - it can be as simple as https://github.com/openshift/machine-config-operator/blob/master/templates/common/_base/units/openvswitch.service.yaml

Alternatively, we can drop in files via the CNO -> MCO.